src/main/java/org/jenkinsci/main/modules/sshd/AsynchronousCommand.java

```diff
@@ -117,7 +116,7 @@ public void run() {
             SecurityContext old = SecurityContextHolder.getContext();
             User user = getCurrentUser();
             if (user!=null)
-                ACL.impersonate(user.impersonate());
+                ACL.as(user);
```

Comment:
This should only be used in a `try`-with-resources block.

Reply:
Do you mean something like this?

```suggestion
                synchronized(this){
                    try{
                        ACL.as(user);
                    }
                }
```

Reply:
No, like

```java
try (ACLContext ctx = ACL.as(user)) {
    // use it
}
```

Reply:
Instead of the `SecurityContextHolder` and `ACL.impersonate` calls. See Javadoc, or just impl of `as`.